src/Workspaces/Core/Portable/Workspace/Workspace_SourceGeneration.cs

```diff
@@ -0,0 +1,132 @@
+        await workspaceStatusService.WaitUntilFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
+
+        await this.SetCurrentSolutionAsync(
+            useAsync: true,
+            oldSolution =>
+            {
+                var updates = GetUpdatedSourceGeneratorVersions(oldSolution, projectIds);
+                return oldSolution.WithSourceGeneratorExecutionVersions(updates, cancellationToken);
+            },
+            static (_, _) => (WorkspaceChangeKind.SolutionChanged, projectId: null, documentId: null),
+            onBeforeUpdate: null,
+            onAfterUpdate: null,
+            cancellationToken).ConfigureAwait(false);
+
+        return;
+
+        static SourceGeneratorExecutionVersionMap GetUpdatedSourceGeneratorVersions(
+            Solution solution, ImmutableSegmentedList<(ProjectId? projectId, bool forceRegeneration)> projectIds)
+        {
+            // For all the projects explicitly requested, update their source generator version.  Do this for all
+            // projects that transitively depend on that project, so that their generators will run as well when next
+            // asked.
+            var dependencyGraph = solution.GetProjectDependencyGraph();
+            var result = SourceGeneratorExecutionVersionMap.CreateBuilder();
+
+            // Determine if we want a major solution change, forcing regeneration of all projects.
+            var solutionMajor = projectIds.Any(t => t.projectId is null && t.forceRegeneration);
+
+            // If it's not a major solution change, then go update the versions for all projects requested.
+            if (!solutionMajor)
+            {
+                // Do a pass where we update minor versions if requested.
+                PopulateSourceGeneratorExecutionVersions(major: false);
+
+                // Then update major versions.  We do this after the minor-version pass so that major version updates
+                // overwrite minor-version updates.
+                PopulateSourceGeneratorExecutionVersions(major: true);
+            }
+
+            // Now, if we've been asked to do an entire solution update, get any projects we didn't already mark, and
+            // update their execution version as well.
+            if (projectIds.Any(t => t.projectId is null))
+            {
+                foreach (var projectId in solution.ProjectIds)
+                {
+                    if (!result.ContainsKey(projectId))
+                    {
+                        result.Add(
+                            projectId,
+                            Increment(solution.GetSourceGeneratorExecutionVersion(projectId), solutionMajor));
+                    }
+                }
+            }
+
+            return result.ToImmutable();
+
+            void PopulateSourceGeneratorExecutionVersions(bool major)
+            {
+                foreach (var (projectId, forceRegeneration) in projectIds)
```

Comment:
>foreach (var (projectId, forceRegeneration) in projectIds) [](http://example.com/codeflow?start=16&length=58)

Could the same projectId be in projectIds multiple times? 

Assuming so, any thoughts on short-circuiting when it's already in result (for the same major value)? Would prevent calls to GetProjectsThatTransitivelyDependOnThisProject which isn't exactly cheap if I remember.

Reply:
> Could the same projectId be in projectIds multiple times?

Yes.  It could be in the list with both 'force=true' and 'force=false'.

> Assuming so, any thoughts on short-circuiting when it's already in result (for the same major value)? Would prevent calls to GetProjectsThatTransitivelyDependOnThisProject which isn't exactly cheap if I remember.

Generally speaking, this runs at a very low pri/cadence.  i'm not very concerned about this walk.

Reply:
So I'm clear here, projectIds is deduped at some point before here? Is that in the ABWQ?

Reply:
the *pair* of projectid+bool is deduped.  But then we process both in this code.  which is why we do the minor-pass first, then the major pass.  that way if there is both, we first do a minor increment, but then overwrite it with a major increment.

Reply:
Yep, that was what I was hoping for. Thanks!